lib/open_project/text_formatting/formats/markdown/pandoc_downloader.rb

```diff
@@ -0,0 +1,87 @@
+#-- encoding: UTF-8
+
+#-- copyright
+# OpenProject is a project management system.
+# Copyright (C) 2012-2018 the OpenProject Foundation (OPF)
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License version 3.
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2017 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See docs/COPYRIGHT.rdoc for more details.
+#++
+
+require 'open3'
+require 'fileutils'
+require 'rest-client'
+
+module OpenProject::TextFormatting::Formats
+  module Markdown
+    module PandocDownloader
+      class << self
+        def check_or_download!
+          return if ENV['OPENPROJECT_PANDOC_PATH'] || executable?
+
+          warn "Downloading recent version of amd64 pandoc for OpenProject ..."
+          download!
+
+          raise "Failed to download pandoc" unless executable?
+        end
+
+        def executable?
+          File.executable?(pandoc_path)
+        end
+
+        def pandoc_path
+          ENV.fetch('OPENPROJECT_PANDOC_PATH', Rails.root.join('vendor/pandoc/bin/pandoc'))
+        end
+
+        def vendored_pandoc_dir
+          Rails.root.join('vendor/pandoc')
+        end
+
+        def pandoc_amd64_tar
+          'https://github.com/jgm/pandoc/releases/download/2.2.3.2/pandoc-2.2.3.2-linux.tar.gz'
+        end
+
+        private
+
+        def download!
+          response = RestClient::Request.execute method: :get,
+                                                 url: pandoc_amd64_tar,
+                                                 raw_response: true
+          tempfile = response.file
```

Comment:
This needs error handling. Ideally the user would be informed that the setup is about to download pandoc and from where (the configured URL). When it fails for what ever reason an error should be shown that the download failed. Maybe a pointer to OPENPROJECT_PANDOC_PATH to allow for manual installation would be good here too. And also one to the var allowing to change the download URL.